design for v2 directory api

jaredkosanovic opened the pull request:
This design versions the directory API to remove the home address and home phone fields, as well as removes the ability to filter results by home phone number. While we're versioning, I also made some other changes to more align the api with JSONAPI spec.

Commit:
return bad request when ldap exception is thrown

it makes more sense to throw a 400 error rather than 500 for caught exceptions. This will help us narrow down 500 responses in appdynamics to uncaught/unexpected errors

jaredkosanovic committed:
Merge pull request #20 from osu-mist/feature/CO-631-500-response

return bad request when ldap exception is thrown

Commit:
clean up, updating

the swagger spec wasn't reflective of what the api actually returns. I updated some things and did some clean up

Commit:
add relationships to link to academic endpoint

jaredkosanovic committed:
Merge pull request #21 from osu-mist/feature/CO-760-design-digital-measures

Feature/co 760 design digital measures

Commit:
use oauth2 api

Commit:
use separate oauth2 url

jaredkosanovic committed:
Merge pull request #22 from osu-mist/feature/CO-786-oauth2-api

Feature/co 786 oauth2 api

jaredkosanovic committed:
update to 1.1.0

jaredkosanovic committed:
Merge pull request #59 from osu-mist/feature/CO-753-update-dw1.1.0

update to 1.1.0

jaredkosanovic committed:
add functions, variables, for MAX_PAGE_SIZE

jaredkosanovic committed:
Merge pull request #60 from osu-mist/feature/co-801-locations-services-bugs

add functions, variables, for MAX_PAGE_SIZE

jaredkosanovic committed:
don't make it final so we can overwrite

jaredkosanovic committed:
Merge pull request #61 from osu-mist/feature/co-801-locations-services-bugs

don't make it final so we can overwrite

jaredkosanovic committed:
use the name of the file that jenkins/api VMs expect

Commit:
Update unicode UID and Search Term vars to valid name

CO-850

Commit:
Merge pull request #23 from osu-mist/feature/CO-850-fix-testGetBySearchQuerySanitizeUnicode

Update unicode UID and Search Term vars to valid name

jaredkosanovic committed:
Merge pull request #62 from osu-mist/feature/CO-867

use the name of the file that jenkins/api VMs expect

magical committed:
ignore page numbers and sizes equal to zero

CO-949

If a user gives a page number or size equal to zero, ignore it and use
the default page number or size.

Page size must not be zero because that can lead to division by zero
errors when we calculate pagination. (CO-949)

Page number must not be zero because we consider it to be 1-indexed, and
going below that doesn't make sense.

magical committed:
update unit test comments

CO-949

These two test cases (negative page number and negative page size) were
fixed ages ago in CO-619. Update the comments to match.

magical committed:
Merge pull request #63 from osu-mist/feature/CO-949-page-size-zero

CO-949: ignore page size of zero

tsoliangwu0130 committed:
Return 400 error for over-borad search term

tsoliangwu0130 committed:
Merge pull request #24 from osu-mist/feature/CO-957-update-integration-test

Return 400 error for over-board search term

jaredkosanovic committed:
update skeleton to dw 1.2.4

jaredkosanovic committed:
Merge pull request #64 from osu-mist/feature/CO-990-update-skeleton

update skeleton to dw 1.2.4

magical committed:
add missing integration tests

CO-1050

Add some tests that were present in DirectoryEntityDAOTest, but not in
integration_tests.py.

Also, make sure we're actually testing what we say we're testing: when
searching by query, the directory api return a 200 response even if
there are 0 results. it's not enough to just test the status code, we
also need to look at the actual result.

Also, fix two tests that had the same method name, and tweak one test to
make the formatting match the other tests.

Also, fix the command line parsing.

magical committed:
add unit tests for sanitize method

CO-1050

magical committed:
remove integration tests from DirectoryEntityDAOTest

CO-1050

These tests are all covered by the Python integration tests now.
Removing them from the groovy test suite allows us to test the
directory-api without having to supply a configuration file
or make any network requests.

magical committed:
refactor integration tests

Make constructing request urls a little nicer

jaredkosanovic committed:
implement additional query parameters

I did a rough clean up of the unit tests so they would pass. Those will eventually be moved to integration tests anyway as apart of CO-1050 (@magical).

jaredkosanovic committed:
clean up swagger, add parameters

Cleaned up the swagger a bit to make it similar to our other swagger files.

jaredkosanovic committed:
Merge remote-tracking branch 'skeleton/master' into feature/CO-1079-update-skeleton

jaredkosanovic committed:
add resource unit tests

jaredkosanovic committed:
make expected/actual order consistent

magical committed:
Update unicode integration test

The account we used to check for seems to be gone.

magical committed:
Merge remote-tracking branch 'origin/feature/CO-1079-add-query-parameters' into feature/CO-1050-bye-bye-integration-tests-rework

jaredkosanovic committed:
Merge pull request #26 from osu-mist/feature/CO-1079-update-skeleton

Feature/co 1079 update skeleton

jaredkosanovic committed:
Merge pull request #25 from osu-mist/feature/CO-1079-add-query-parameters

implement additional query parameters

magical committed:
Merge pull request #27 from osu-mist/feature/CO-1050-bye-bye-integration-tests

CO-1050: move directory api integration tests out of dropwizard

jaredkosanovic committed:
add mockito dependency

jaredkosanovic committed:
make mockito verison explicit

jaredkosanovic committed:
Merge pull request #28 from osu-mist/feature/mockito-fix

add mockito dependency

jaredkosanovic committed:
remove relationships

This block was added for the digital measures ticket. We can always add it back, but in an effort to make the swagger reflect what the API produces, let's remove it for now.

jaredkosanovic committed:
Merge pull request #29 from osu-mist/remove-relationships

remove relationships

jaredkosanovic committed:
add accepted response builder

jaredkosanovic committed:
replace fullname with first and last, remove leading wildcards

jaredkosanovic committed:
Merge pull request #30 from osu-mist/feature/CO-1106-remove-leading-wildcard

replace fullname with first and last, remove leading wildcards

jaredkosanovic committed:
Merge pull request #65 from osu-mist/feature/CO-1082-add-accepted-response

add accepted response builder

tsoliangwu0130 committed:
Update reademe

tsoliangwu0130 committed:
Update reademe

tsoliangwu0130 committed:
Update reademe

tsoliangwu0130 committed:
Merge pull request #66 from osu-mist/feature/CO-460-readme-update

Update reademe

jaredkosanovic committed:
update ojdbc driver

jaredkosanovic committed:
Merge pull request #67 from osu-mist/bug/CO-1145-encrypt-oracle-db-connections

update ojdbc driver

jaredkosanovic committed:
add forbidden response

this code originally came from the osu-events-api

jaredkosanovic committed:
add noContent (204) response

Commit:
Refactor ldap connection pool logic out of the DirectoryDAO

Commit:
build and runtime information should be in a meta member

Commit:
Clean up imports in the DirectoryEntityDAO

Commit:
Remove unused var in DirectoryEntityDAO.groovy

Commit:
Fix missing return type to configureLdapPool

Commit:
Fix line spacing in configureLdapPool

Commit:
Change def to full type and split across multiple lines

Commit:
Merge pull request #31 from osu-mist/refactor/CO-294-ldap-connection-pool

Refactor ldap connection pool logic out of the DirectoryDAO

wilsonia committed:
create InfoResultObject for wrapping meta information

wilsonia committed:
create InfoResultObject for wrapping meta information

wilsonia committed:
resolve merge conflicts

wilsonia committed:
create ErrorResultObject to wrap errors in a top-level array

wilsonia committed:
error fields match JSON spec

wilsonia committed:
bring all error objects up to JSON spec

wilsonia committed:
bring link object up to JSON spec

wilsonia committed:
update documentation

wilsonia committed:
default detail values

wilsonia committed:
update version in example curl request

jaredkosanovic committed:
Merge pull request #68 from osu-mist/feature/CO-1163-test-message-api

Add forbidden (403) and no content (204) responses

wilsonia committed:
remove unnecessary default empty message values

wilsonia committed:
add forbidden error

wilsonia committed:
Merge branch 'master' into feature/CO-1136-follow-JSON-spec

wilsonia committed:
update version

wilsonia committed:
Merge pull request #70 from osu-mist/feature/CO-1136-follow-JSON-spec

Feature/co 1136 follow json spec

wilsonia committed:
forbidden error in error collection

wilsonia committed:
default value for parameter so calls made without don't fail

wilsonia committed:
Merge pull request #72 from osu-mist/bug/CO-1136-builders-need-default-argument

Bug/co 1136 builders need default argument

tsoliangwu0130 committed:
Add auth header

tsoliangwu0130 committed:
Add auth header

tsoliangwu0130 committed:
Modify description

tsoliangwu0130 committed:
Modify description

tsoliangwu0130 committed:
Merge pull request #73 from osu-mist/feature/CO-1229-add-auth-headeer

CO-1229 Add auth header

tsoliangwu0130 committed:
Merge pull request #32 from osu-mist/feature/CO-1229-add-auth-headeer

CO-1229 Add auth header

tsoliangwu0130 committed:
Add host, operationId

tsoliangwu0130 committed:
Merge pull request #33 from osu-mist/feature/CO-1276-update-dev-portal

CO-1276 Add host, operationId

wilsonia committed:
add compound error response builder

wilsonia committed:
change argument to ErrorResultObject list

wilsonia committed:
move wrapping of error object in ErrorResultObject to Resource

wilsonia committed:
document change in type for error objects

wilsonia committed:
Merge pull request #74 from osu-mist/feature/CO-1238-Compound-Error-Builder

add compound error response builder

jaredkosanovic committed:
design for v2 directory api

This design versions the directory API to remove the home address and home phone fields, as well as removes the ability to filter results by home phone number. While we're versioning, I also made some other changes to more align the api with JSONAPI spec.

jaredkosanovic committed:
merge from skeleton

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -1,8 +1,8 @@
-  version: "1"
+  title: Directory API
+  description: Web API to search the OSU Online Directory.
+  version: "2"
```
We decided to implement the new version of this API with our new express-skeleton, so the version should be using `v2` since it will be used as part of the code. Also, can you change the file name to `openapi.yaml`?

jaredkosanovic replied:
I'll make it v2, but how will you document minor version changes? Does v2.0 make more sense?

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -12,7 +12,7 @@ externalDocs:
 host: api.oregonstate.edu
```
Can you also add `contact` field? I mainly wanna keep all APIs consistent by following the same template: https://github.com/osu-mist/express-api-skeleton/blob/master/openapi.yaml#L9-L12

jaredkosanovic replied:
Done

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -94,9 +89,7 @@ paths:
-            type: array
-            items:
-              $ref: "#/definitions/ResultObject"
+            $ref: "#/definitions/ResultObjects"
```
Can this be renamed as `DirectoryResults` or something more specific? Also, I don't think the `pretty` parameters is needed anymore, can you remove it as well?

jaredkosanovic replied:
Done

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -197,21 +190,26 @@ definitions:
       self:
         type: string
   Error:
```
Could you use `Errors` (plural) like [here](https://github.com/osu-mist/express-api-skeleton/blob/master/openapi.yaml#L188-L217)? You can simply copy the `Error` and `Errors` objects from the express-api-skeleton.

jaredkosanovic replied:
Done

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -139,16 +132,20 @@ parameters:
+    properties:
+      data:
+        $ref: "#/definitions/ResourceObject"
   ResourceObject:
```
For `ResourceObject`, `ResultObject` and `ResultObjects`, I'd suggest rename them more specificly.

jaredkosanovic replied:
Done

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -197,21 +190,26 @@ definitions:
       self:
         type: string
```
Can you also add `format: url` to this field?

jaredkosanovic replied:
Done

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -139,16 +132,20 @@ parameters:
+        type: string
       type:
         type: string
         example: "directory"
```
Instead of using `example`, can you use `enum: [directory]` instead?

jaredkosanovic replied:
Done

tsoliangwu0130 reviewed (changes requested):
Some other comments:

1. Sometimes the string was quoted by single-quote, sometimes by double-quotes, and sometimes it isn't quoted at all. Can you make them more consistent and only use single-quote if necessary?
2. I'm not sure how valuable it'll be to allow user search with `osuuid` for the `getByID` endpoint. In my opinion, search by OSU ID makes more sense. What do you think?

jaredkosanovic committed:
changes from code review

jaredkosanovic committed:
rename swagger -> openapi

jaredkosanovic committed:
delete swagger.json

old file that was forgotten and neglected, so I put it out of its misery

jaredkosanovic committed:
change osuuid parameter and attribute to string

The osuuid was an int in v1, but since it was used in the 'id', that violated jsonAPI spec since it should be a string instead. I made the ID string in a previous commit, but I forgot about the parameter and redudant osuuid field in the attributes which we can probably remove

jaredkosanovic commented:
> 2\. I'm not sure how valuable it'll be to allow user search with `osuuid` for the `getByID` endpoint. In my opinion, search by OSU ID makes more sense. What do you think?

This API is currently getting data from ldap which uses the osuuid as the primary identifier. I don't think (public) ldap has the 9 digit OSU ID, but even if it did, the 9 digit OSU ID comes from banner and there might be some people that exist in ldap who don't exist in banner.

What ID do your API clients want? Just curious, since I've never figured out what the best ID to use.

tsoliangwu0130 commented:
> This API is currently getting data from ldap which uses the osuuid as the primary identifier. I don't think (public) ldap has the 9 digit OSU ID, but even if it did, the 9 digit OSU ID comes from banner and there might be some people that exist in ldap who don't exist in banner. What ID do your API clients want? Just curious, since I've never figured out what the best ID to use.

Good questions. The reason I feel like `UID` is not really useful is because I got some requests from our clients recently to ask the relationship between `OSU UID` and `OSU ID`. Most of them doesn't know what `UID` is, so they may also need to get the permission of identify API to convert the `OSU ID` to `OSU UID`. I'll bring this up during the team meeting to see if others have any thought on it. Good work!

wilsonia commented:
These changes all make sense to me. After discussion with @tsoliangwu0130 and @fernmig we have decided to use UID as the top level ID, and Miguel is looking into using OSU ID as another returned field / parameter. I will hold off on changing ID fields until we know what we can do.

This PR will be merged into master, as the default branch is currently develop. When v2 is implemented, the branch used for deployment will change from develop to master.
